### DISCUSSION TITLE
I hope GODOT will add the most critical 2D function in 4.0, that is, IK and particle collision that can be controlled by code

### DISCUSSION THREAD

**Discussion by RIKDXHQ11** [godotengine/godot-proposals#5823] [2D]
I hope GODOT will add the most critical 2D function in 4.0, that is, IK and particle collision that can be controlled by code
In the 4.0 alpha version, I still haven't found any IK nodes that can be controlled by the code. That is to say, in the horizontal version of the game, when a character is aiming with a gun, and the crosshair is close to his arm, his wrist can't be bent. The characters can only extend their arms clumsily.
The IK provided with the engine can only be used for animation, and cannot be used in code.
There is also a particle system that does not collide. Stars and liquids cannot fall on the floor surface. It penetrates the ground. And then disappeared

**Comment by Calinou** [comment]
> There is also a particle system that does not collide. Stars and liquids cannot fall on the floor surface. It penetrates the ground. And then disappeared

GPUParticles collision is already implemented in 4.0.beta, but you can't control particles on the CPU for performance reasons. There's an option to hide particles on contact still, which works as long as you've created colliders that match your scene's visuals.

**Reply by RIKDXHQ11** [reply]
> > There is also a particle system that does not collide. Stars and liquids cannot fall on the floor surface. It penetrates the ground. And then disappeared
> 
> GPUParticles collision is already implemented in 4.0.beta, but you can't control particles on the CPU for performance reasons. There's an option to hide particles on contact still, which works as long as you've created colliders that match your scene's visuals.

This is great, I can use it.
The remaining function is that IK can be controlled by the code, which is necessary, because if it is implemented, it will greatly help the animation of real-time calculation (this technology is used in many 3D games, such as characters opening doors, holding various types of weapons, grabbing objects, etc., without pre animation)
